lib/Query/QueryBuilderConverter.php

```diff
@@ -0,0 +1,251 @@
+     * @var Metadata[]
+     */
+    protected $documentMetadata = [];
+
+    /**
+     * @param PropertyEncoder
+     */
+    protected $encoder;
+
+    /**
+     * @param DocumentStrategyInterface
+     */
+    private $strategy;
+
+    /**
+     * @param SessionInterface $session
+     * @param EventDispatcherInterface $eventDispatcher
+     * @param MetadataFactoryInterface $metadataFactory
+     * @param PropertyEncoder $encoder
+     */
+    public function __construct(
+        SessionInterface $session,
+        EventDispatcherInterface $eventDispatcher,
+        MetadataFactoryInterface $metadataFactory,
+        PropertyEncoder $encoder,
+        DocumentStrategyInterface $strategy
+    ) {
+        $this->eventDispatcher = $eventDispatcher;
+        $this->qomf = $session->getWorkspace()->getQueryManager()->getQOMFactory();
+        $this->metadataFactory = $metadataFactory;
+        $this->encoder = $encoder;
+        $this->strategy = $strategy;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getQuery(QueryBuilder $builder)
+    {
+        $this->documentMetadata = [];
+        $this->sourceDocumentNodes = [];
+        $this->constraint = null;
+
+        $this->locale = $builder->getLocale();
+
+        if (!$this->locale) {
+            throw new \InvalidArgumentException(sprintf(
+                'No locale specified'
+            ));
+        }
+
+        $from = $builder->getChildrenOfType(QBConstants::NT_FROM);
+
+        if (!$from) {
+            throw new RuntimeException(
+                'No From (source) node in query'
+            );
+        }
+
+        $dispatches = [
+            QBConstants::NT_FROM,
+            QBConstants::NT_SELECT,
+            QBConstants::NT_WHERE,
+            QBConstants::NT_ORDER_BY,
+        ];
+
+        foreach ($dispatches as $dispatchType) {
+            $this->dispatchMany($builder->getChildrenOfType($dispatchType));
+        }
+
+        if (count($this->sourceDocumentNodes) > 1 && null === $builder->getPrimaryAlias()) {
+            throw new \InvalidArgumentException(
+                'You must specify a primary alias when selecting from multiple document sources ' .
+                'e.g. $qb->from(\'a\') ...'
+            );
+        }
+
+        $this->applySourceConstraints($builder);
+
+        $phpcrQuery = $this->qomf->createQuery(
+            $this->from,
+            $this->constraint,
+            $this->orderings,
+            $this->columns
+        );
+
+        $query = new Query($phpcrQuery, $this->eventDispatcher, $this->locale, [], $builder->getPrimaryAlias());
+
+        if ($firstResult = $builder->getFirstResult()) {
+            $query->setFirstResult($firstResult);
+        }
+
+        if ($maxResults = $builder->getMaxResults()) {
+            $query->setMaxResults($maxResults);
+        }
+
+        return $query;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applySourceConstraints(QueryBuilder $builder)
```

Comment:
It also looks like this method is not inherited, so the `{@inheritdoc}` doesn't make any sense.